File: unflatten.js
#!/usr/bin/env node
const fs = require('fs');

function unflattenNodeData(dataArray) {
  const base = {};
  if (dataArray.length < 2 || typeof dataArray[1] !== 'object' || Array.isArray(dataArray[1])) return base;
  if (typeof dataArray[0] === 'object' && !Array.isArray(dataArray[0])) Object.assign(base, dataArray[0]);
  const mapping = dataArray[1], pool = dataArray, memo = {}, inProgress = new Set();
  function resolve(idx) {
    if (typeof idx !== 'number' || idx < 0 || idx >= pool.length) return null;
    if (memo.hasOwnProperty(idx)) return memo[idx];
    if (inProgress.has(idx)) return pool[idx];
    inProgress.add(idx);
    const val = pool[idx];
    let result;
    if (val && typeof val === 'object' && !Array.isArray(val) &&
        Object.values(val).every(v => typeof v === 'number')) {
      result = {};
      for (const [k, subIdx] of Object.entries(val)) result[k] = resolve(subIdx);
    } else if (Array.isArray(val)) {
      result = val.map(x => typeof x === 'number' ? resolve(x) : x);
    } else result = val;
    memo[idx] = result;
    inProgress.delete(idx);
    return result;
  }
  for (const [key, idx] of Object.entries(mapping)) base[key] = resolve(idx);
  return base;
}

function unflattenNodes(nodes) {
  return (nodes || [])
    .filter(raw => raw && raw.type === 'data')
    .map(raw => unflattenNodeData(raw.data));
}

const inputPath = process.argv[2];
const root = JSON.parse(fs.readFileSync(inputPath, 'utf-8'));
const records = unflattenNodes(root.nodes);
const main = records.find(r => r && r.uuid);
const final = main || (records.length ? records[0] : {});
fs.writeFileSync(inputPath, JSON.stringify(final, null, 2));

File: unflatten.py
#!/usr/bin/env python3
import json, sys

def unflatten_node_data(data_array):
    base = {}
    if len(data_array) < 2 or not isinstance(data_array[1], dict):
        return base

    if isinstance(data_array[0], dict):
        base.update(data_array[0])

    mapping = data_array[1]
    pool = data_array
    memo = {}
    in_progress = set()

    def resolve(idx):
        if not isinstance(idx, int) or idx < 0 or idx >= len(pool):
            return None
        if idx in memo:
            return memo[idx]
        if idx in in_progress:
            return pool[idx]
        in_progress.add(idx)
        val = pool[idx]
        if isinstance(val, dict) and val and all(isinstance(v, int) for v in val.values()):
            result = {k: resolve(v) for k, v in val.items()}
        elif isinstance(val, list):
            result = [resolve(x) if isinstance(x, int) else x for x in val]
        else:
            result = val
        memo[idx] = result
        in_progress.remove(idx)
        return result

    for key, idx in mapping.items():
        base[key] = resolve(idx)
    return base


def unflatten_nodes(nodes):
    out = []
    for raw in nodes or []:
        if isinstance(raw, dict) and raw.get("type") == "data":
            arr = raw.get("data")
            if isinstance(arr, list):
                out.append(unflatten_node_data(arr))
    return out

if __name__ == "__main__":
    input_path = sys.argv[1]
    root = json.load(open(input_path))
    records = unflatten_nodes(root.get("nodes"))
    main = next((r for r in records if isinstance(r, dict) and 'uuid' in r), None)
    final = main if main is not None else (records[0] if records else {})
    open(input_path, "w").write(json.dumps(final, indent=2))
